src/main/java/edu/kit/compiler/semantic/DetailedNameTypeAstVisitor.java

```diff
@@ -0,0 +1,491 @@
+
+        if (objectType.getType() != DataTypeClass.UserDefined) {
+            semanticError(fieldAccessExpressionNode, "field access is only allowed on reference type expressions");
+        }
+
+        namespace = namespaceMapper.getClassNamespace(objectType.getIdentifier().get());
+
+        if (!namespace.getClassSymbols().containsKey(fieldAccessExpressionNode.getName())) {
+            semanticError(fieldAccessExpressionNode, "unknown class field");
+        }
+        ClassNodeField definition = namespace.getClassSymbols().get(fieldAccessExpressionNode.getName());
+
+        fieldAccessExpressionNode.setDefinition(definition);
+
+        fieldAccessExpressionNode.setResultType(definition.getType());
+        return definition.getType();
+    }
+
+    public DataType visit(ArrayAccessExpressionNode arrayAccessExpressionNode) {
+        DataType objectType = arrayAccessExpressionNode.getObject().accept(this);
+        if (objectType.getType() != DataTypeClass.Array) {
+            semanticError(arrayAccessExpressionNode, "%s is not an array type", objectType.getRepresentation(stringTable));
+        }
+
+        DataType expressionType = arrayAccessExpressionNode.getExpression().accept(this);
+        if (expressionType.getType() != DataTypeClass.Int) {
+            semanticError(arrayAccessExpressionNode, "array index must be of type int");
+        }
+
+        DataType resultType = objectType.getInnerType().get();
+
+        arrayAccessExpressionNode.setResultType(resultType);
+        return resultType;
+    }
+
+    public DataType visit(IdentifierExpressionNode identifierExpressionNode) {
+        if (!symboltable.isDefined(identifierExpressionNode.getIdentifier())) {
+            semanticError(identifierExpressionNode, "`%s` cannot be resolved", stringTable.retrieve(identifierExpressionNode.getIdentifier()));
+        }
+        Definition definition = symboltable.lookup(identifierExpressionNode.getIdentifier());
+
+        identifierExpressionNode.setDefinition(definition);
+
+        identifierExpressionNode.setResultType(definition.getType());
+        return definition.getType();
+    }
+
+    public DataType visit(ThisExpressionNode thisExpressionNode) {
+        if (!currentClassNamespace.isPresent()) {
+            semanticError(thisExpressionNode, "`this` is not allowed in static contexts");
+        }
+
+        ClassNode classNode = currentClassNamespace.get().getClassNodeRef();
+        DataType resultType = new DataType(classNode.getName());
+
+        thisExpressionNode.setDefinition(classNode);
+
+        thisExpressionNode.setResultType(resultType);
+        return resultType;
+    }
+
+    public DataType visit(ValueExpressionNode valueExpressionNode) {
+        DataType resultType;
+        switch (valueExpressionNode.getValueType()) {
+        case False:
+            resultType = new DataType(DataTypeClass.Boolean);
+            break;
+        case IntegerLiteral:
```

Comment:
We need to ensure that integer literals actually fit into a 32-bit signed integer. So you need something like the following:
```
try {
    valueExpressionNode.literalValue.get().intValue();
} catch (NumberFormatException e) {
    // ...
}
```

Reply:
Good point, I thought we already checked that elsewhere.